### PR TITLE
Migrate DatasetBasedTimeScheduleStore to the new SPI

### PR DESCRIPTION
**Pull request by jayfjin**
Migrate DatasetBasedTimeScheduleStore to the new SPI. The SQL test will be added when the JDBC work is done - since then it will be easy to add the sql test by passing in the appropriate cconf. 

One known flaky test failing, otherwise green: https://builds.cask.co/browse/CDAP-DUT6841-1/test

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 1]
```diff
@@ -661,4 +664,30 @@ public static void createTable(StructuredTableAdmin tableAdmin) throws IOExcepti
       tableAdmin.create(PROGRAM_LINEAGE_SPEC);
     }
   }
+
+  /**
+   *
```
Comment (CuriousVini) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: javadoc?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStoreTest.java [hunk 2]
```diff
@@ -359,19 +352,7 @@ public void testTriggersWithoutJobs() throws Exception {
...
-    factory.createExecutor(ImmutableList.of((TransactionAware) table))
-      .execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() throws Exception {
-          datasetBasedTimeScheduleStore.removeJob(table, job1.getKey());
-        }
-      });
-
-    // Make sure Jobs and Triggers are setup correctly
-    verifyJobAndTriggers(job1.getKey(), 1, Trigger.TriggerState.NORMAL);
```
Comment (CuriousVini) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStoreTest.java:373]: why is this removed?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java [hunk 3]
```diff
@@ -48,30 +52,29 @@
...
+import javax.annotation.Nullable;
 
 /**
  * ScheduleStore extends from RAMJobStore and persists the trigger and schedule information into datasets.
  */
 public class DatasetBasedTimeScheduleStore extends RAMJobStore {
   private static final Logger LOG = LoggerFactory.getLogger(DatasetBasedTimeScheduleStore.class);
-  private static final byte[] JOB_KEY = Bytes.toBytes("jobs");
-  private static final byte[] TRIGGER_KEY = Bytes.toBytes("trigger");
+  private static final String JOB_KEY = "job";
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:68]: should this be `jobs`?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStoreTest.java (hunk 2 above)
Comment (jayfjin) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStoreTest.java:373]: It's done above. With the new implementation, it doesn't make sense to check after I remove job1 - it'll already be gone by then

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java (hunk 3 above)
Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:68]: no - the type is a job or a trigger

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java (hunk 1 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: Done

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java [hunk 4]
```diff
@@ -204,108 +207,109 @@ private void persistJobAndTrigger(final JobDetail newJob, final OperableTrigger
...
-    byte[][] col = new byte[1][];
-    col[0] = Bytes.toBytes(key.getName());
-    Row result = table.get(TRIGGER_KEY, col);
-    byte[] bytes = null;
-    if (!result.isEmpty()) {
-      bytes = result.get(col[0]);
-    }
-    if (bytes != null) {
-      return (TriggerStatusV2) SerializationUtils.deserialize(bytes);
+  TriggerStatusV2 readTrigger(StructuredTable table, TriggerKey key) throws IOException {
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:254]: This should be marked as nullable

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java [hunk 5]
```diff
@@ -204,108 +207,109 @@ private void persistJobAndTrigger(final JobDetail newJob, final OperableTrigger
...
-  @VisibleForTesting
-  void removeJob(Table table, JobKey key) {
-    byte[][] col = new byte[1][];
-    col[0] = Bytes.toBytes(key.toString());
-    table.delete(JOB_KEY, col);
+
+  @Nullable
+  private byte[] get(StructuredTable table, String type, String name) throws IOException {
+    Optional<StructuredRow> row = table.read(getPrimaryKeys(type, name));
+    if (!row.isPresent()) {
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:239]: would this be better `row.map(structuredRow -> structuredRow.getBytes(StoreDefinition.TimeScheduleStore.VALUE_FIELD)).orElse(null);`?

**Review by CuriousVini — APPROVED** [review]
2 minor comments. Rest lgtm. Please squash and merge after DUT and ITM is green.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java (hunk 5 above)
Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:239]: imo prefer not to map when we're operating on a singleton

**Comment by jayfjin** [comment]
ITN green https://builds.cask.co/browse/IT-ITM-106
